### PR TITLE
Adding LenientJsonArraySerializer

### LINKED ISSUE #9: `ListSerializer` that ignores invalid elements
https://github.com/Kotlin/kotlinx.serialization/issues/1205

> **What is your use-case and why do you need this feature?** I'm receiving a list of items from the server. Some of the items might miss some required fields or be otherwise malformed. In this case, I would like to ignore the items that failed to parse but still parse the ones that I can. Unfortunately, it looks like `ListSerializer` is failing as soon as it encounters a single list item which it fails to parse.
> 
> **Describe the solution you'd like** Some sort of `ignoreFailedListItems` or `allowPartialListResults` configuration.
> 
> *This probably applies to all collections.

### PR DESCRIPTION
**Pull request by Kantis**
Fixes #9 

Not sure if there was something new in KxS 1.6 that made this possible, or if I just ended up trying something different.. but it seems to work 🎉

### CODE REVIEW

File: ks3-standard/src/commonMain/kotlin/io/ks3/standard/ResilientListSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import kotlinx.serialization.serializer
+
+inline fun <reified T> resilientListSerializer(explicitElementSerializer: KSerializer<T>? = null) =
+   object : KSerializer<List<T>> {
+      private val elementSerializer = explicitElementSerializer ?: serializer<T>()
+      private val serializer: SerializationStrategy<List<T>> = ListSerializer(elementSerializer)
+      override val descriptor: SerialDescriptor = serializer.descriptor
+
+      override fun deserialize(decoder: Decoder): List<T> {
+         if (decoder !is JsonDecoder) error("This deserializer can only be used with Json")
```
Comment (aSemy) [inline ks3-standard/src/commonMain/kotlin/io/ks3/standard/ResilientListSerializer.kt:20]:
```suggestion
         require(encoder is JsonEncoder) { "resilientListSerializer can only be used with Json" }
```

A soft suggestion.

I've got a slight preference for `require()`, and `IllegalArgumentException` is a slightly better fit than `IllegalStateException`.

File: ks3-standard/src/commonMain/kotlin/io/ks3/standard/ResilientListSerializer.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+         val json = decoder.json
+         val list = decoder.decodeJsonElement()
+         return list
+            .jsonArray
+            .mapNotNull {
+               runCatching {
+                  json.decodeFromJsonElement(elementSerializer, it)
+               }.getOrNull()
```
Comment (aSemy) [inline ks3-standard/src/commonMain/kotlin/io/ks3/standard/ResilientListSerializer.kt:28]:
```suggestion
               try {
                  Json.decodeFromJsonElement(elementSerializer, it)
               } catch (ex: IllegalArgumentException) {
                  null // ignore failures
               }
```

I think it's better to try and only catch the exception that is documented by `decodeFromJsonElement()`

File: ks3-standard/src/commonMain/kotlin/io/ks3/standard/ResilientListSerializer.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            }
+      }
+
+      override fun serialize(
+         encoder: Encoder,
+         value: List<T>,
+      ) = serializer.serialize(encoder, value)
```
Comment (aSemy) [inline ks3-standard/src/commonMain/kotlin/io/ks3/standard/ResilientListSerializer.kt:35]:
```suggestion
      ): T = serializer.serialize(encoder, value)
```

My preference is to explicitly state the return type for expression functions, just so it's clearer when reading the source code, and it prevents any changes in the return type from 'leaking' up stream.

Maybe that's something Ktlint can handle? :)

**Comment by aSemy** [comment]
Nice!

File: ks3-standard/src/commonMain/kotlin/io/ks3/standard/ResilientListSerializer.kt (hunk 1 above)
Comment (Kantis) [reply ks3-standard/src/commonMain/kotlin/io/ks3/standard/ResilientListSerializer.kt:20]: Thanks, totally agree.. I try to use those whenever I remember :)

**Comment by Kantis** [comment]
what do you feel about the name? 
1. `resilientListSerializer`  what I came up with when trying this out again
2. `listSerializerIgnoringInvalidItems`  seen in the issue

I'm kind of in favor of no 2.

**Comment by aSemy** [comment]
Yeah, any of those names is fine. Personally I'd go with `LenientJsonArraySerializer`. I'll leave it up to you :)

**Review by aSemy — APPROVED** [review]

**Comment by Kantis** [comment]
Thanks, that's a great suggestion :) I'll go with that

**Pull request merged**